[Security Solution] Add cypress tests to entity flyouts

machadoum opened the pull request:
## Summary

issue: https://github.com/elastic/kibana/pull/173392

Flaky test runner runs: 
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4773
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4841


* Add cypress test to the new User and Host flyouts.
* Fix function using the wrong host risk score index
* Fix some cypress flakiness

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

machadoum committed:
Add cypress test to entity flyout

machadoum committed:
REVERT ME

machadoum committed:
Ooops, forgot to commit some files

machadoum committed:
Fix bug that fetched the wrong risk score index

machadoum committed:
Make cypress test more reliable by using ES archiver

machadoum commented on x-pack/plugins/security_solution/common/search_strategy/security_solution/risk_score/common/index.ts:
```diff
@@ -24,7 +24,9 @@ export const getHostRiskIndex = (
   isNewRiskScoreModuleInstalled: boolean
 ): string => {
   return isNewRiskScoreModuleInstalled
-    ? getRiskScoreLatestIndex(spaceId)
```
I found this bug that we didn't consider the `onlyLatest` property for the host index. I fixed it by copying the user index code.

semd replied:
https://github.com/elastic/kibana/blob/64ea66731bee85806f28e19de881deaa2e2e5842/STYLEGUIDE.mdx#L401-L414

machadoum commented on x-pack/plugins/security_solution/public/flyout/shared/components/expandable_panel.tsx:
```diff
@@ -88,7 +88,7 @@ export const ExpandablePanel: React.FC<ExpandablePanelPanelProps> = ({
     expandedOnFirstRender: false,
   },
-  'data-test-subj': dataTestSubj,
+  'data-test-subj': dataTestSubj = 'expandablePanel',
```
`data-test-subj` is optional, so it is better to define a default value instead of rendering `undefined`.

PhilippeOberti replied:
@machadoum 

hmmmm I'm actually not sure this is better. While I understand that having undefined isn't a really good solution either, having a default value on a reusable component means that we will end up with potentially many many components in the DOM with the same `data-test-subj`. Yes is was the case with `undefined`, but now the value actually exists and developers might be tempted to use it, which could end up with multiple hits.

Following good practices, `data-test-subj` values should always be unique.

Would there be a way to conditionally set the property if it's value exists instead?

PhilippeOberti replied:
if it's not easy to do the condition, I would revert that change and keep `undefined` for now until we find a better solution. We already have those optional `data-test-subj` in a bunch of places, so it would make sense to tackle all of them at the same time with a good fix...
Thoughts?

machadoum replied:
Ok. I will revert the change.

machadoum commented on x-pack/test/security_solution_cypress/cypress/tasks/asset_criticality/common.ts:
```diff
@@ -26,6 +28,7 @@ export const expandFirstAlertHostFlyout = () => {
  * Find the first alert row in the alerts table then click on the host name to open the flyout
  */
 export const expandFirstAlertUserFlyout = () => {
+  scrollAlertTableColumnIntoView(USER_COLUMN);
```
It didn't click when the column was not visible.

tiansivive replied:
AHHHH! So this must be why the tests I disabled were also flaky

machadoum commented on x-pack/test/security_solution_cypress/cypress/tasks/entity_analytics.ts:
```diff
@@ -42,6 +42,7 @@ export const navigateToNextPage = () => {
 export const riskEngineStatusChange = () => {
   cy.get(RISK_SCORE_SWITCH).should('not.have.attr', 'disabled');
   cy.get(RISK_SCORE_SWITCH).click();
+  cy.get(RISK_SCORE_STATUS).should('have.text', 'On');
```
Wait for the enable risk engine response to arrive and make sure it is enabled before proceeding. It fixed the test flakiness 🥳

hop-dev replied:
Nice! I was looking at something similar for one of our other flaky tests I wonder if this is the fix

machadoum committed:
Revert "REVERT ME"

This reverts commit bd47f5428c9a7156e03318403872f93c8e1ad0b9.

machadoum commented on x-pack/test/security_solution_cypress/es_archives/risk_scores_new/data.json:
```diff
@@ -483,3 +483,504 @@
 }
+
+
+{
```
Add the same data to the data stream index.

rylnd replied:
What data is this the same as? And why is it being added?

machadoum replied:
Previously, the file only inserted documents into the latest index, which resulted in the time-series index being empty during tests. However, after my change, it will now insert the same documents into both indices. So, any features that rely on the time-series index will now function as expected.

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-entity-analytics (Team:Entity Analytics)

machadoum committed:
Merge branch 'main' into siem-ea-8084-cypress

hop-dev reviewed (approved):
Great work! 🚀

machadoum committed:
Revert undefined data-test-subj fix

machadoum committed:
Fix cypress test

machadoum committed:
Merge remote-tracking branch 'elastic/main' into siem-ea-8084-cypress

machadoum committed:
Update risk score cypress tests to enable risk engine before loading risk data

The risk engine enablement fails if there is data already inside the risk score index

machadoum commented on x-pack/test/security_solution_cypress/cypress/e2e/entity_analytics/dashboards/entity_analytics.cy.ts:
```diff
@@ -330,21 +330,15 @@ describe('Entity Analytics Dashboard', { tags: ['@ess', '@serverless'] }, () =>
 
     describe('Risk Score enabled but still no data', () => {
       before(() => {
-        cy.task('esArchiverLoad', { archiveName: 'risk_scores_new_no_data' });
```
I noticed that many risk engine tests are showing an error when enabling the engine because the index was already created. So, I am updating the tests to enable the risk engine before loading data with the archiver.

machadoum committed:
Revert changes to risk score archiver and create a new archiver for entity flyout

machadoum committed:
Revert cypress changes

machadoum committed:
Simplify ternary operator

machadoum committed:
Merge remote-tracking branch 'elastic/main' into siem-ea-8084-cypress

machadoum committed:
Run entity flyout test on serverless

machadoum committed:
Fix host risk tab cypress test

machadoum commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into siem-ea-8084-cypress

semd commented on x-pack/plugins/security_solution/common/search_strategy/security_solution/risk_score/common/index.ts:
```diff
@@ -23,21 +23,23 @@ export const getHostRiskIndex = (
-      ? getRiskScoreLatestIndex(spaceId)
-      : getRiskScoreTimeSeriesIndex(spaceId)
-    : `${RISKY_USERS_INDEX_PREFIX}${onlyLatest ? 'latest_' : ''}${spaceId}`;
+  if (isNewRiskScoreModuleInstalled) {
+    return onlyLatest ? getRiskScoreLatestIndex(spaceId) : getRiskScoreTimeSeriesIndex(spaceId);
+  } else {
+    return `${RISKY_USERS_INDEX_PREFIX}${onlyLatest ? 'latest_' : ''}${spaceId}`;
+  }
```
NIT: this `getUserRiskIndex` implementation is 99% the same as `getHostRiskIndex`. Maybe it's worth extracting this logic into a `getEntityRiskIndex`, receiving the index prefix (the only thing that changes), this way we keep the code DRY.

semd commented on x-pack/test/security_solution_cypress/cypress/e2e/entity_analytics/new_entity_flyout.cy.ts:
```diff
@@ -0,0 +1,87 @@
+  () => {
+    before(() => {
+      login();
+      enableRiskEngine();
```
I think that Glo mentioned that the only thing we should do in the _before_ and _after_ hooks is loading archives, anything else (like `enableRiskEngine` or `deleteRiskEngineConfiguration` calls) should go to _beforeEach_ and _afterEach_, is this correct @MadameSheema?

MadameSheema replied:
correct @semd

machadoum replied:
Unfortunately, we cannot load the archived data without enabling the risk engine first. 

The only thing I can do is move everything to `beforeEach` and `afterEach,` which will slow down the test execution and could create flakiness. The problem here is that we don't have a way to delete every resource that is created when we enable the engine (we should have that). So when we enable the engine for the second time, it keeps the previously created data, and that could impact the test.

Could you please share the documentation for this `before` and `after` rule? In this case, how can we execute some logic only once inside Cypress tests?

machadoum replied:
@MadameSheema @semd I decided to change the test approach. Now, instead of enabling the risk engine, I am mocking the risk engine status API. That will make the test more reliable. I also fixed some archiver bugs that prevented the test from running twice.

What do you think?

https://github.com/elastic/kibana/pull/174232/commits/edd34473ae47a3655d94429d1326053b1293ad3d#diff-145ec61fabe13c482eb0a6a759f06b3455b61578d4b32135f3c1363bccbae2a9L62-L69

MadameSheema replied:
> Could you please share the documentation for this before and after rule? 

Hey @machadoum!! I'm working on more extensive documentation on automation best practices, but you can find that statement [here](https://github.com/elastic/kibana/issues/173327) and [here](https://github.com/elastic/kibana/issues/174247). 

> In this case, how can we execute some logic only once inside Cypress tests?

Can you be more specific about cases where you want to execute a logic just once?

> @MadameSheema @semd I decided to change the test approach. Now, instead of enabling the risk engine, I am mocking the risk engine status API. That will make the test more reliable. I also fixed some archiver bugs that prevented the test from running twice.

What do you think?

As the scope of the test is not testing the process of enabling the risk engine, I believe is a great idea towards having more reliable and stable tests :)

semd replied:
@machadoum @MadameSheema 
That's a good temporary solution, it will prevent flakiness. However, I think it would be better if we were able to test the real API, to make sure we are testing the UI -> API integration property. 

Nevertheless, from what I know, and correct me if I am wrong, installing the risk engine is not a quick task, it may take some minutes. So, installing it in the beforeEach and removing it in the afterEach test case is not the best approach. 

What do you think if we unblock this PR and merge it with the API mock, and also create a separate issue to implement a helper/util library to install the risk engine in the _before_ hook and clean at the end in the _after_ hook?

machadoum replied:
It works for me. 👍 

I brought to the team's attention the need for an API that completely deletes all risk engine assets. If we have this API we could call it once inside the `after` block.

semd reviewed (commented):
Everything looks good Pablo, I only need to confirm the _before_ and _after_ hooks question.

machadoum committed:
Improve cypress tests by mocking the risk engine status API

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/e2e/entity_analytics/new_entity_flyout.cy.ts:
```diff
@@ -0,0 +1,86 @@
+
+    describe('User details', () => {
+      it('should display entity flyout and open risk input panel', () => {
+        visitWithTimeRange(ALERTS_URL);
```
NIT: This and below line can be moved to the beforeEach hook.

MadameSheema commented:
@machadoum please exercise the new introduced changes on the flaky test suite runner before merging in order to make sure we are not introducing flakiness.

machadoum committed:
Revert query_alerts change because it breaks serverless cypress

semd reviewed (approved):
LGTM

machadoum committed:
Update query_alerts unload to allow multiple test runs

machadoum committed:
Implement nit code improvement

MadameSheema reviewed (approved):
Thanks for all the improvements! :)
